[presentation-api] Update tests for onconnect, onclose and onterminate

tomoyukilabs opened the pull request:
This PR upgrades #4330 and fixes #4040 as well as several additional issues.

* `async_test` and `step_func` are replaced with `promise_test` and `EventWatcher`.
* Each file name is corrected and `.https` is added to the file name.
* The `onclose` and `onterminate` tests check behavior in each connection state: `connecting`, `connected`, `closed` and `terminated`.
* The `onterminate` test also checks behavior of another presentation connection in a nested browsing context.
* The `onterminate` test now supposes that calling `connection.terminate()` must set state of presentation connections in `connecting` and `connected` state to `terminated` as defined in the current spec: https://github.com/w3c/presentation-api/pull/427

<!-- Reviewable:start -->

<!-- Reviewable:end -->

tomoyukilabs committed:
update tests for onconnect, onclose and onterminate

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/229034890)

# Firefox (nightly channel) #
Testing web-platform-tests at revision ad7f2f216d5d98f044446b0df903debfc079ecba
Using browser at version BuildID 20170504100357; SourceStamp 33b92d9c40562dab3d7b602368c75619f1d793f7
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>4 tests ran</summary>
<details>
<summary><a href="https://w3c-test.org/submissions/5794/presentation-api/controlling-ua/PresentationRequest_sandboxing_error.https.html">/presentation-api/controlling-ua/PresentationRequest_sandboxing_error.https.html</a></summary>


|                                                   Subtest                                                   | Results |                                                       Messages                                                       |
|-------------------------------------------------------------------------------------------------------------|---------|----------------------------------------------------------------------------------------------------------------------|
|                                                                                                             | OK      |                                                                                                                      |
| `Sandboxing: Creating a PresentationRequest from a nested context fails when allow-presentation is not set` | FAIL    | `assert_equals: Presentation sandboxing did not work as expected. expected "SecurityError" but got "ReferenceError"` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/5794/presentation-api/controlling-ua/PresentationRequest_sandboxing_success.https.html">/presentation-api/controlling-ua/PresentationRequest_sandboxing_success.https.html</a></summary>


|                                                  Subtest                                                   | Results |                                                    Messages                                                    |
|------------------------------------------------------------------------------------------------------------|---------|----------------------------------------------------------------------------------------------------------------|
|                                                                                                            | OK      |                                                                                                                |
| `Sandboxing: Creating a PresentationRequest from a nested context succeeds when allow-presentation is set` | FAIL    | `assert_equals: Presentation sandboxing did not work as expected. expected "success" but got "ReferenceError"` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/5794/presentation-api/controlling-ua/getAvailability_sandboxing_success.https.html">/presentation-api/controlling-ua/getAvailability_sandboxing_success.https.html</a></summary>


|                                                   Subtest                                                   | Results |                                                               Messages                                                               |
|-------------------------------------------------------------------------------------------------------------|---------|--------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                             | OK      |                                                                                                                                      |
| `Sandboxing: Retrieving display availability from a nested context succeeds when allow-presentation is set` | FAIL    | `assert_equals: Presentation sandboxing did not work as expected. expected "success" but got "Could not create PresentationRequest"` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/5794/presentation-api/controlling-ua/reconnectToPresentation_sandboxing_success.https.html">/presentation-api/controlling-ua/reconnectToPresentation_sandboxing_success.https.html</a></summary>


|                                                 Subtest                                                 | Results |                                                                  Messages                                                                  |
|---------------------------------------------------------------------------------------------------------|---------|--------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                         | OK      |                                                                                                                                            |
| `Sandboxing: Reconnecting a presentation from a nested context succeeds when allow-presentation is set` | FAIL    | `assert_equals: Presentation sandboxing did not work as expected. expected "NotFoundError" but got "Could not create PresentationRequest"` |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/229034896)

# Chrome (unstable channel) #
Testing web-platform-tests at revision ad7f2f216d5d98f044446b0df903debfc079ecba

tomoyukilabs committed:
remove trailing whitespace

tidoust reviewed (changes requested):
I see a couple of errors to fix in the "onclose" test. Please see inline.

tomoyukilabs commented on presentation-api/controlling-ua/PresentationConnection_onclose-manual.https.html:
```diff
@@ -0,0 +1,128 @@
+      connection = c;
+      return stash.init();
+    }).then(() => {
+      // Step 1: close the presentation connection in "connecting" state
```
Oops, that's right. I'll move `stash.init()` somewhere else.

By the way, while I'm not sure whether it is correct or not, Chrome Canary throws an exception like `"InvalidAccessError: PresentationRequest::start() requires user gesture."` when `request.start()` is called within `stash.init().then()`. This is the reason why I avoided calling `stash.init()` before `request.start()`.

tidoust replied:
I don't know. I believe the behavior is somewhat up to implementations in practice, see end of third bullet point in definition of [allowed to show a popup](http://www.w3.org/TR/html51/browsers.html#allowed-to-show-a-popup).

It's interesting that `Promise.all` seems to solve the problem.

tomoyukilabs commented on presentation-api/controlling-ua/PresentationConnection_onclose-manual.https.html:
```diff
@@ -0,0 +1,128 @@
+      eventWatcher = new EventWatcher(t, connection, ['close', 'connect', 'terminate']);
+
+      connection.close();
+      return Promise.all([
```
Oops, I'll fix it.

tomoyukilabs committed:
Fix a couple of bugs

- `stash.init()` is called before `request.start()`
- inappropriate use of `Promise.all` is changed to `Promise.race'

tomoyukilabs committed:
check an event passed by both onxxx and addEventListener

tomoyukilabs commented:
I have fixed a couple of bugs in the "onclose" test. Also, I have updated the three tests so that they check both `on(connect|close|terminate)` and addEventListener (used in `EventWatcher`).
